File: Week_09/reverse_only_letters/Solution.java
package reverse_only_letters;

import java.util.Stack;

/**
 * 子母栈 O(n)，O(n)
 */
class Solution1 {
    public String reverseOnlyLetters(String S) {
        Stack<Character> stack = new Stack<>();
        for (char c : S.toCharArray()) {
            if (Character.isLetter(c))
                stack.push(c);
        }
        StringBuilder res = new StringBuilder();
        for (char c : S.toCharArray()) {
            if (Character.isLetter(c))
                res.append(stack.pop());
            else
                res.append(c);
        }
        return res.toString();
    }
}

/**
 * 双指针 O(n)，O(n)
 */
class Solution2 {
    public String reverseOnlyLetters(String S) {
        char[] c = S.toCharArray();
        int i = 0, j = c.length - 1;
        while (i < j) {
            if (!Character.isLetter(c[i])) i++;
            else if (!Character.isLetter(c[j])) j--;
            else { char tmp = c[i]; c[i++] = c[j]; c[j--] = tmp; }
        }
        return new String(c);
    }
}

File: Week_09/reverse_only_letters/Solution.py
# Two pointer, reverse char array but skip char that is not a letter
class Solution1:
    def reverseOnlyLetters(self, S: str) -> str:
        s = list(S) # python str is immutable, convert to list first
        i, j = 0, len(s) - 1
        while i < j:
            while i < j and not s[i].isalpha(): i += 1
            while i < j and not s[j].isalpha(): j -= 1
            if i < j:
                s[i], s[j] = s[j], s[i]
                i, j = i + 1, j - 1
        return "".join(s)


# Stack
class Solution2:
    def reverseOnlyLetters(self, S: str) -> str:
        letters = [c for c in S if c.isalpha()]

        ans = []
        for c in S:
            if c.isalpha():
                ans.append(letters.pop())
            else:
                ans.append(c)

        return "".join(ans)